physical/raft/transaction.go

```diff
@@ -0,0 +1,605 @@
+}
+
+func doVerifyEntry(key string, value []byte, expected []byte) error {
+	if len(expected) < 1 {
+		return fmt.Errorf("truncated verification hash: %w", physical.ErrTransactionCommitFailure)
+	}
+
+	hashType := expected[0]
+
+	actual, err := createVerificationEntryOfType(hashType, key, value)
+	if err != nil {
+		return err
+	}
+
+	if subtle.ConstantTimeCompare(actual, expected) == 0 {
+		return fmt.Errorf("for get %v (actual=%v vs expected=%v - actual value `%v`): %w", key, hex.EncodeToString(actual), hex.EncodeToString(expected), hex.EncodeToString(value), physical.ErrTransactionCommitFailure)
+		// return physical.ErrTransactionCommitFailure
+	}
+
+	return nil
+}
+
+func parseListVerifyParams(repr string) (*verifyListOpParams, error) {
+	var params verifyListOpParams
+	if err := json.Unmarshal([]byte(repr), &params); err != nil {
+		return nil, err
+	}
+
+	return &params, nil
+}
+
+func doVerifyList(repr string, items []string, expected []byte) error {
+	hashType := expected[0]
+
+	sItems := strings.Join(items, "\n")
+
+	actual, err := createVerificationEntryOfType(hashType, repr, []byte(sItems))
+	if err != nil {
+		return err
+	}
+
+	if subtle.ConstantTimeCompare(actual, expected) == 0 {
+		return fmt.Errorf("for list %v: %w", repr, physical.ErrTransactionCommitFailure)
+		// return physical.ErrTransactionCommitFailure
+	}
+
+	return nil
+}
+
+func cloneBytes(val []byte) []byte {
+	ret := make([]byte, len(val))
+	copy(ret, val)
+	return ret
+}
+
+type raftTxnUpdateRecord struct {
+	// If this record exists but Contents is nil, the entry was deleted.
+	Contents *physical.Entry
+}
+
+type RaftTransaction struct {
+	b              *RaftBackend
+	l              sync.RWMutex
+	tx             *bbolt.Tx
+	updates        map[string]*raftTxnUpdateRecord
+	log            *LogData
+	writable       bool
+	haveWritten    bool
+	haveFinishedTx bool
+}
+
+var _ physical.Transaction = &RaftTransaction{}
+
+func (b *RaftBackend) newTransaction(ctx context.Context, writable bool) (*RaftTransaction, error) {
+	// Grab a transaction permit pool entry so that we can limit the number of
+	// concurrent transactions. Also grab a read lock in the underlying FSM
+	// to prevent key changes from occurring while a transaction is ongoing.
+	// These will be released when we finish this transaction.
+	b.txnPermitPool.Acquire()
+	b.fsm.l.RLock()
+
+	// All transactions are read-only; this gives us a consistent view of
+	// storage but means we need to track writes ourselves.
+	tx, err := b.fsm.db.Begin(false)
+	if err != nil {
+		return nil, fmt.Errorf("failed to start underlying bbolt transaction: %w", err)
+	}
+
+	return &RaftTransaction{
+		b:       b,
+		tx:      tx,
+		updates: make(map[string]*raftTxnUpdateRecord),
+		log: &LogData{
+			Operations: []*LogOperation{
+				{
+					OpType: beginTxOp,
+				},
+			},
+		},
+		writable: writable,
+	}, nil
+}
+
+func (t *RaftTransaction) Put(ctx context.Context, entry *physical.Entry) error {
+	if !t.writable {
```

Comment:
I guess this doesn't strictly need to be an atomic since it can't be modified, but it'd be good practice to make this consistent.

Reply:
I moved these checks inside the lock. We'll need the lock anyways, so there's no point in using an atomic for this as erring early doesn't buy us much.